### PR TITLE
DrivAerML: probe LR above 5e-4 under EMA+gc champion (5.5e-4, 6e-4)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DrivAerML LR optimum under the EMA+gc champion config (EMA=0.9995, gc=0.5) may be **above** 5e-4, not at it. Evidence from PR #3158 shows a steep, monotonic degradation *below* lr=5e-4:

- lr=4.5e-4: +7.9% worse than baseline
- lr=4e-4: +54.5% worse than baseline

This asymmetric LR basin strongly suggests the loss landscape has more headroom above 5e-4 that has never been probed. The grad-clip at 0.5 may be providing enough stability to tolerate a modestly higher peak LR.

## Instructions

Run two full trials under the champion config, changing only `--lr`:

**Trial 1 — lr=5.5e-4 (conservative upward step — run this first)**

Start with a 3-epoch smoke test. Report `val_primary/surface_rel_l2_pct` after epoch 3. Target: beat 3.833% (baseline) or be meaningfully on track. If the smoke test looks healthy, let it run to convergence (epochs=999, SENPAI_MAX_EPOCHS=9999).

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5.5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb_group dm-higher-lr
```

**Trial 2 — lr=6e-4 (aggressive upward step)**

Run Trial 2 in parallel (or immediately after Trial 1's smoke test passes). Do NOT wait for Trial 1 to fully converge before starting Trial 2.

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 6e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb_group dm-higher-lr
```

**Stop early if:** either run diverges (val loss spikes >2x baseline by epoch 5). Report best `val_primary/surface_rel_l2_pct` for each trial at convergence.

## Baseline

Current best metrics (DrivAerML, PR #3072 — eren/dm-ema-9995-gc05):
- **val_primary/surface_rel_l2_pct: 3.833%** (epoch 511)
- test surface_rel_l2_pct: 4.685%
- Baseline W&B run: `ncl1dh88` (eren/dm-ema-9995-gc05)
- External target: <3.71% (AB-UPT, gap = 0.123 pp, 1.03x)

Reproduce baseline:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT bulma:

## Results

### Trial 1 — lr=5.5e-4 (conservative upward step)

| Metric | Baseline (lr=5e-4) | Trial 1 (lr=5.5e-4) | Delta |
|---|---|---|---|
| val_primary/surface_rel_l2_pct | **3.833%** (ep 511) | 4.390% (ep 279) | +0.557 pp (+14.5%) |
| test surface_rel_l2_pct | 4.685% | 5.264% | +0.579 pp (+12.4%) |
| External target | <3.71% (AB-UPT) | — | — |

- **W&B run:** [`c15pv0g0`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/c15pv0g0)
- **Total epochs:** 437 (timeout at 360 min)
- **Diverged:** Yes — grad_norm → Infinity after ~epoch 420. Val metric collapsed to ~59% in late epochs. EMA checkpoint at epoch 279 preserved the best result.
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5.5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 \
  --agent bulma --wandb-name "bulma/dm-lr-5.5e-4" --wandb-group dm-higher-lr
```

### Trial 2 — lr=6e-4 (aggressive upward step)

| Metric | Baseline (lr=5e-4) | Trial 2 (lr=6e-4) | Delta |
|---|---|---|---|
| val_primary/surface_rel_l2_pct | **3.833%** (ep 511) | 4.227% (ep 426) | +0.394 pp (+10.3%) |
| test surface_rel_l2_pct | 4.685% | 4.843% | +0.158 pp (+3.4%) |
| External target | <3.71% (AB-UPT) | — | — |

- **W&B run:** [`kqadbq1x`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/kqadbq1x)
- **Total epochs:** 427 (timeout at 360 min)
- **Diverged:** No — training remained stable throughout. Was still improving slightly at timeout (best at ep 426 of 427).
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=3 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 6e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 \
  --agent bulma --wandb-name "bulma/dm-lr-6e-4" --wandb-group dm-higher-lr
```

## What happened

**Both trials are clearly worse than baseline.** The hypothesis that the LR optimum lies above 5e-4 is not supported:

- lr=5.5e-4 was 14.5% worse on val and eventually diverged (grad_norm → ∞ after ~epoch 420)
- lr=6e-4 was 10.3% worse on val but remained stable and was still improving at timeout

Counterintuitively, the more aggressive lr=6e-4 outperformed the conservative lr=5.5e-4. This may be because the cosine restarts (t-max=30) interact differently — at lr=6e-4 the restarts create sharper exploration that helps escape local minima, while lr=5.5e-4 sits in a less favorable basin.

However, neither comes close to the baseline 3.833%. Combined with the prior data from PR #3158 (lr=4.5e-4 was +7.9% worse, lr=4e-4 was +54.5% worse), this confirms that **lr=5e-4 sits at a sharp optimum** for the DrivAerML champion config. The LR basin is narrow and asymmetric — degradation is steep in both directions.

## Bug fix

Fixed an `UnboundLocalError` in `train.py` where a local variable `primary_metric_key` (lines 1645-1648) was shadowing the function `primary_metric_key()` (line 1428), causing a crash on line 1639. Removed the redundant local assignment and used `best_val_primary_metric_name` instead.

## Suggested follow-ups

- The narrow LR basin suggests the optimizer dynamics (cosine restarts + grad clip) may be the limiting factor, not the peak LR itself. Could explore warmup-based restarts or exponential decay instead of hard cosine restarts.
- lr=6e-4 was still improving at timeout — if given more epochs it might close the gap further, though likely won't beat baseline.
- The sharp LR sensitivity points to a loss landscape issue — adaptive methods like schedule-free AdamW or gradient noise injection might widen the basin.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3194 because ADVISOR: Closing — lr=5.5e-4 (4.390%) and lr=6e-4 (4.227%) both worse than 3.700% baseline. lr=5.5e-4 diverged ep~420; lr=6e-4 stable but 14% worse. Confirms lr=5e-4 is the champion LR. Bulma reassigned..
